aiida_lsmo/workchains/cp2k_phonopy.py

```diff
@@ -0,0 +1,188 @@
+
+
+class Cp2kPhonopyWorkChain(WorkChain):
+    """A workchain to compute frequencies using CP2K and Phonopy"""
+
+    @classmethod
+    def define(cls, spec):
+        """Define workflow specification."""
+        super().define(spec)
+
+        spec.input('structure', valid_type=(CifData, StructureData), required=True, help='Input structure')
+        spec.expose_inputs(Cp2kBaseWorkChain,
+                           namespace='cp2k_base',
+                           exclude=['cp2k.structure', 'cp2k.parameters', 'cp2k.metadata.options.parser_name'])
+
+        spec.outline(
+            cls.generate_displacements,
+            cls.collect_cp2k_inputs,
+            cls.run_cp2k_first,
+            while_(cls.should_run_displ)(cls.run_cp2k_displacements,),
+            cls.results,
+        )
+
+        # Ouptuts
+
+    def generate_displacements(self):
+        """Generate displacements using Phonopy"""
+
+        # CifData to PhonopyAtoms
+        ase = self.inputs.structure.get_ase()
+        uc = PhonopyAtoms(symbols=ase.get_chemical_symbols(),
+                          cell=ase.get_cell(),
+                          scaled_positions=ase.get_scaled_positions())
+
+        # Generate displacements
+        self.ctx.phonon = Phonopy(
+            unitcell=uc,
+            supercell_matrix=None,
+            primitive_matrix=None,
+            factor=CP2KToTHz,
+        )
+        self.ctx.phonon.generate_displacements(distance=0.01)
+        disp_ucs = self.ctx.phonon.supercells_with_displacements
+        #phonon.save() to investigate
+
+        # List of PhonopyAtoms to list of StructureData
+        self.ctx.sd_list = []
+        for i, pa in enumerate([uc] + disp_ucs):  # 6N+1 in total, first is the original uc
+            sd = StructureData(cell=uc.get_cell())
+            for i in range(len(uc)):
+                symbol = uc.get_chemical_symbols()[i]
+                position = uc.get_positions()[i]
+                sd.append_atom(symbols=symbol, position=position)
+            self.ctx.sd_list.append(sd)
+
+    def collect_cp2k_inputs(self):
+        """Collect Cp2k inputs from the last CP2K calculation."""
+        qb = QueryBuilder()
+        qb.append(Node, filters={f'id': self.inputs.structure.pk}, tag='cif_out')
+        qb.append(Node, filters={'attributes.process_label': 'Cp2kCalculation'}, with_descendants='cif_out', tag='calc')
+        last_cp2k_calc = qb.distinct().all()[-1][0]
+
+        self.ctx.base_inp = AttributeDict(self.exposed_inputs(Cp2kBaseWorkChain, 'cp2k_base'))
+        self.ctx.base_inp['cp2k']['settings'] = Dict(dict={'additional_retrieve_list': ['aiida-forces-1_0.xyz']})
+        self.ctx.base_inp['cp2k']['file'] = {}
+
+        for edge_label in last_cp2k_calc.inputs:
+            if edge_label.startswith('file'):
+                edge_label_file = edge_label.split('__')[-1]
+                self.ctx.base_inp['cp2k']['file'][edge_label_file] = last_cp2k_calc.inputs[edge_label]
+
+        param_modify = Dict(
```

Comment:
I think this is one of the least important lacks of provenance in this wc: I don't care much about inputs to a calculation, since practically you rarely need to check their exact provenance.